### PR TITLE
SAK-43480 - Filtering wrong decimal separator characters in the new G…

### PR DESCRIPTION
**Pull request by danielmerino**
…rader respecting to user's locale

### CODE REVIEW

**Comment by danielmerino** [comment]
Codacy is rejecting a change with this error: "Found non-literal argument to RegExp Constructor".

I need that rule to be disabled, because the regexp argument is dynamically built. AFAIK there's no other option but disabling it, at least in this patch.

**Comment by ern** [comment]
Would using [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat) be ideal?

**Comment by danielmerino** [comment]
> Would using [Intl.NumberFormat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat) be ideal?

Sorry, Earle, I don't understand. Intl.NumberFormat would be useful for parsing a number, but that number is what user is entering in the grade field.

The case we are detecting here is when some user with an specific locale is typing the grade using a decimal separator wrong for his locale. Having that the wrong decimal separator is a valid thousand separator in each locale, I can't imagine how NumberFormat could help.

The existing code filters all non-valid characters typed in the field. I have just added the correct decimal separator to the accepted characters.

File: webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js [hunk 1]
```diff
@@ -800,9 +800,19 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
...
+      i18nDecimalPattern = "[\d.]";
+    }
+    else {
+      i18nDecimalPattern = "[\d,]";
+    }
+    var rgxp = new RegExp(i18nDecimalPattern);
+
     if (e.key === "Backspace" || e.key === "ArrowLeft" || e.key === "ArrowRight") {
       return true;
-    } else if (!e.key.match(/[\d.,]/)) {
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:805]: I know I'm a little late to this, but don't we just need a + after the closing ]?
/[\d.,]+/

All your change seems to be doing is making the decimal character more specific, based on the portal locale, whereas we should just be able to accept all of the decimal chars.

**Comment by adrianfish** [comment]
Try doing this in gradeSelected

this.submission.grade = this.submission.grade.replace(",",".");

File: webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js [hunk 2]
```diff
@@ -800,9 +800,19 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
...
 
+    var i18nDecimalPattern = "";
+    /*global portal*/
+    if (this.getDecimalSeparator(portal.locale) === ".") {
+      i18nDecimalPattern = "[\d.]";
+    }
+    else {
+      i18nDecimalPattern = "[\d,]";
+    }
+    var rgxp = new RegExp(i18nDecimalPattern);
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:811]: const decimalSeparator = (1.1).toLocaleString(portal.locale).substring(1, 2);
const rgxp = new RegExp(`[\\d${decimalSeparator}]`);

File: webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js [hunk 3]
```diff
@@ -817,6 +827,15 @@ export class SakaiGrader extends gradableDataMixin(SakaiElement) {
...
     }
   }
+  
+  getDecimalSeparator(locale) {
+    const numberWithDecimalSeparator = 1.1;
+
+    return numberWithDecimalSeparator
+        .toLocaleString(locale)
+        .substring(1, 2);
+  }
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/grader/sakai-grader.js:837]: You don't need this function

**Review by adrianfish — CHANGES_REQUESTED** [review]
I get your change now! You're just blocking any decimal point/comma input that's not valid for the current locale. Nice. I've just made your change a bit more compact.

**Comment by danielmerino** [comment]
Hi @adrianfish unfortunately I'm going on holidays today and I won't be able to do your requested changes until middle of July.

If you want to change the commit and merge it, I don't mind if you give to yourself the authory. Otherwise, I will do it when I return. Thanks.

**Comment by adrianfish** [comment]
Hi @danielmerino No worries, I'll do that. I'll add a commit so we'll both be on it. Thanks for the change, and enjoy your break. I hope you do something nice :)

**Review by adrianfish — APPROVED** [review]

**Pull request merged**